tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+//!         if let Ok(read) = http3_conn.recv_body(&mut conn, stream_id, &mut buf)
+//!         {
+//!             test.add_response_body(stream_id, &buf, read);
+//!         }
+//!     },
+//!
+//!     _ => ()
+//! }
+//! # Ok::<(), quiche::h3::Error>(())
+//! ```
+//!
+//! ## Tests assertion
+//!
+//! Once all responses have been received and their data stored
+//! the [`assert()`] method can be used to validate that everthing was ok.
+//!
+//! Calling [`assert()`] prematurely will likely result in failure so it helps to
+//! track requests as so:
+//!
+//! ```no_run
+//! # let server = "https://cloudflare-quic.com/b";
+//! # let mut url = url::Url::parse(server).unwrap();
+//! # let mut reqs = Vec::new();
+//!
+//! # let path = format!("{}/{}", url.path(), "get");
+//! # url.set_path(&path);
+//!
+//! # let expect_hdrs = Some(vec![quiche::h3::Header::new(":status", "200")]);
+//! # reqs.push(httpbintest::HttpBinReq::new("GET", &url, None, expect_hdrs));
+//!
+//! # // Using a closure...
+//! # let assert = |reqs: &[httpbintest::HttpBinReq]| {
+//! #   reqs[0].assert_hdrs()
+//! # };
+//!
+//! # let mut test = httpbintest::HttpBinTest::new(url, reqs, assert);
+//! # let mut config = quiche::Config::new(quiche::PROTOCOL_VERSION).unwrap();
+//! # let scid = [0xba; 16];
+//! # let mut conn = quiche::connect(None, &scid, &mut config).unwrap();
+//! # let h3_config = quiche::h3::Config::new(0, 1024, 0, 0)?;
+//! # let mut http3_conn = quiche::h3::Connection::with_transport(&mut conn, &h3_config)?;
+//! let mut requests_complete = 0;
+//! let request_count = test.requests_count();
+//! match http3_conn.poll(&mut conn) {
+//!     Ok((_stream_id, quiche::h3::Event::Finished)) => {
+//!         requests_complete += 1;
+//!         if requests_complete == request_count {
+//!             if !test.assert() {
+//!                 // test validation failed, handle it
+//!             }
+//!         }
+//!     },
+//!     _ => ()
+//! }
+//! # Ok::<(), quiche::h3::Error>(())
+//! ```
+//!
+//! [quiche]: https://github.com/cloudflare/quiche/
+//! [test]: struct.HttpBinTest.html
+//! [`HttpBinTest`]: struct.HttpBinTest.html
+//! [`assert_hdrs()`]: struct.HttpBinTest.html#method.assert_hdrs
+//! [`requests_count()`]: struct.HttpBinTest.html#method.requests_count
+//! [`set_concurrency()`]: struct.HttpBinTest.html#method.set_concurrency
+//! [`add_response_headers()`]: struct.HttpBinTest.html#method.add_response_headers
+//! [`add_response_body()`]: struct.HttpBinTest.html#method.add_response_body
+//! [`assert()`]: struct.HttpBinTest.html#method.assert
+
+#[macro_use]
+extern crate log;
+
+use std::collections::HashMap;
+
+use quiche::h3::Header;
+
+const USER_AGENT: &str = "quiche-http3bin";
+
+// Stores the request, the expected response headers, and the actual response.
+// The assert_hdrs() method is provided for convenience to validate the received
+// headers match the expected headers.
+#[derive(Clone)]
+pub struct HttpBinReq {
```

Comment:
Same regarding mentioning httpbin, maybe just call `HttpTestReq`? This also applies to the other types below.